Add white-list group filter to wavefield writer.

krenzland opened the pull request:
See documentation:

 Similar to the previous parameter, OutputGroups can be used to whitelist a set of
mesh groups (as specified in the xdmf mesh file) that are included in the wavefieldoutput.
Cells whose group is not mentioned are not included in the output.
This feature works with OutputRegionBounds, only cells that satisfy both criteria are included.
It looks like this:

`   OutputGroups = 1 2 ! only include groups 1 and 2`


Note that this is a very old commit from 2020 that I just found lying around - not sure why I never merged this. I just cherry picked this from an old branch and merged with master. The feature was tested then; I haven't run extensive tests now but a small scale test works. The code is very simple as well and just piggy-backs on the old region filter, so I see no reason why it would work on more complicated setups as well.

Thomas-Ulrich commented on Documentation/wave-field-output.rst:
```diff
@@ -57,14 +57,27 @@ OutputRegionBounds
+------------------
+
+Similar to the previous parameter, OutputGroups can be used to whitelist a set of
+mesh groups (as specified in the xdmf mesh file) that are included in the wavefieldoutput.
```
wavefield output

Thomas-Ulrich commented on src/Numerical_aux/typesdef.f90:
```diff
@@ -1091,6 +1091,7 @@ MODULE TypesDef
                                                                                 !< .FALSE. = do not integrate and output for this variable
      REAL                         ,POINTER  :: OutputRegionBounds(:)            !< Region for which the output should be written
                                                                                 !< Format is xMin, xMax, yMin, yMax, zMin, zMax
+     INTEGER, ALLOCATABLE :: OutputGroups(:) !< Stores an array of group ids that should be included in output. All others are excluded.
```
in the volume output.

Thomas-Ulrich commented on src/Reader/readpar.f90:
```diff
@@ -2772,7 +2775,9 @@ SUBROUTINE readpar_output(EQN,DISC,IO,MPI,CalledFromStructCode)
-      ALLOCATE(IO%IntegrationMask(9),STAT=allocstat )                        !
+    IO%OutputGroups = pack(OutputGroups, OutputGroups >= 0)
+
+	  ALLOCATE(IO%IntegrationMask(9),STAT=allocstat )                        !
```
fix code alignment

Thomas-Ulrich replied:
Also don't you need to allocate IO%OutputGroups first?

krenzland replied:
I don't think so - pack seems to allocate. All examples that I can find where pack is used on allocateable arrays do not use allocate manually.

Thomas-Ulrich replied:
ok you're right (I was too lazy to google).

Thomas-Ulrich commented on src/ResultWriter/WaveFieldWriter.cpp:
```diff
@@ -252,28 +260,32 @@ void seissol::writer::WaveFieldWriter::init(unsigned int numVars,
-		for (std::map<int,int>::iterator it=oldToNewVertexMap.begin(); it!=oldToNewVertexMap.end(); ++it)
-			subVertices.at(it->second) = &allVertices[it->first];
+		for (auto & it : oldToNewVertexMap) {
+            subVertices.at(it.second) = &allVertices[it.first];
```
fix alignment? (tabulations...)

Thomas-Ulrich commented on src/ResultWriter/WaveFieldWriter.cpp:
```diff
@@ -202,38 +205,43 @@ void seissol::writer::WaveFieldWriter::init(unsigned int numVars,
-		// Extract elements based on the region specified
+		// Extract elements based on the specified group or region
 		for (size_t i = 0; i < numElems; i++) {
+		    // Note: For some reason, an alternative name for groupId is material.
```
should we change that?

Thomas-Ulrich commented on src/ResultWriter/WaveFieldWriter.cpp:
```diff
@@ -202,38 +205,43 @@ void seissol::writer::WaveFieldWriter::init(unsigned int numVars,
-		/** Extract the elements and vertices based on user given bounds */
-		// Reference to the vector containing all the elements
-		const std::vector<Element>& allElements = meshReader.getElements();
+	const bool extractBox = outputRegionBounds[0] != 0.0
```
extractBoxEnabled and extractGroupEnabled?

Thomas-Ulrich commented on src/ResultWriter/WaveFieldWriter.cpp:
```diff
@@ -202,38 +205,43 @@ void seissol::writer::WaveFieldWriter::init(unsigned int numVars,
+			        || vertexInBox(outputRegionBounds, allVertices[allElements[i].vertices[1]].coords)
+			        || vertexInBox(outputRegionBounds, allVertices[allElements[i].vertices[2]].coords)
+			        || vertexInBox(outputRegionBounds, allVertices[allElements[i].vertices[3]].coords);
+			const bool isInGroup = !extractGroup || outputGroups.count(groupId) > 0; // TODO(Lukas) Implement this.
```
why the comment?

Thomas-Ulrich commented on src/ResultWriter/WaveFieldWriter.cpp:
```diff
@@ -252,28 +260,32 @@ void seissol::writer::WaveFieldWriter::init(unsigned int numVars,
 			oldToNewVertexMap, *tetRefiner);
-	} else {
-		meshRefiner = new refinement::MeshRefiner<double>(meshReader, *tetRefiner);
+    } else {
```
fix alignment? (tabulations...)

Thomas-Ulrich commented on src/ResultWriter/WaveFieldWriter.cpp:
```diff
@@ -252,28 +260,32 @@ void seissol::writer::WaveFieldWriter::init(unsigned int numVars,
 	}
 
+    if (numElems == 0) {
+        logError() << "All cells are excluded in the WaveFieldWriter.";
```
WaveFieldWriter: All cells have been filtered out (OutputRegionBounds and OutputGroups).

Thomas-Ulrich commented:
Just did a quick test (with both extract options). It works:
![image](https://user-images.githubusercontent.com/13536910/161519939-56acaf4e-b531-422d-8e74-85631c6a4dce.png)

krenzland commented:
Thanks, Thomas!

Thomas-Ulrich commented on src/ResultWriter/WaveFieldWriter.cpp:
```diff
@@ -208,21 +208,22 @@ void seissol::writer::WaveFieldWriter::init(unsigned int numVars,
+                                   || outputRegionBounds[5] != 0.0;
+
+  // If at least one group is explicitly enabled, extract
+  const bool isExtractGroupEnabled = !outputGroups.empty();
```
alignment not fixed (check in GitHub).

Thomas-Ulrich commented on src/Geometry/NetcdfReader.h:
```diff
@@ -258,7 +258,7 @@ class NetcdfReader : public MeshReader
 		ElemNeighborRanks* elemNeighborRanks = new ElemNeighborRanks[maxSize];
 		ElemMPIIndices* elemMPIIndices = new ElemMPIIndices[maxSize];
-		ElemMaterial* elemMaterial = new ElemMaterial[maxSize];
+		ElemGroup* elemMaterial = new ElemGroup[maxSize];
```
Have you considered also changing also variable names (e.g. elemMaterial)?

Thomas-Ulrich commented on src/Geometry/PUMLReader.cpp:
```diff
@@ -381,7 +381,7 @@ void seissol::PUMLReader::getMesh(const PUML::TETPUML &puml)
 		}
 
-		m_elements[i].material = material[i];
+		m_elements[i].group = material[i];
```
material --> group?

Thomas-Ulrich reviewed (commented):
Thx, but alignments are not yet fixed...

krenzland commented:
I've fixed the alignment. The entire file was broken because it randomly mixes tabs and spaces. I just replaced all tabs with spaces. Larger diff, but otherwise it's impossible to get the alignment working :( (It was broken before my PR)

Thomas-Ulrich commented:
by the way, at this point you could clang-format the file (using format prescribed in dr-cpp branch).
